src/urllib3/_backends/sync_backend.py

```diff
@@ -49,17 +49,17 @@ def _wait(self, readable, writable, read_timeout=None):
                 timeout=read_timeout):
             raise socket.timeout()  # XX use a backend-agnostic exception
 
-    def receive_some(self):
+    def receive_some(self, read_timeout):
         while True:
             try:
                 return self._sock.recv(BUFSIZE)
             except util.SSLWantReadError:
-                self._wait(readable=True, writable=False)
+                self._wait(readable=True, writable=False, read_timeout=read_timeout)
```

Comment:
Not actually related to this PR, but we should probably rename the `wait` argument to just `timeout`. It has nothing to do with reads in particular...

Reply:
Just did that, thanks! That highlights that read_timeout also applies to write timeouts, which is not what we want, I think?

Reply:
Good question. I just spent 5 minutes looking, and ... I'm not actually sure *how* urllib3's current timeout policy works... this is super confusing.

In `ConnectionPool.urlopen`, the connect timeout does get passed to `_get_conn` (which eventually passes it to `Queue.get`, I think?) and is also assigned to `self.timeout`, which `HTTPConnection._new_conn` checks and eventually passes through to `socket.create_connection`. So I guess we end up with a total connect timeout that's twice the specific connect timeout. And then later, after the request is made, `connectionpool.py` checks the read timeout, and uses it to set the socket timeout. But in between those two times, we send the actual request. And AFAICT, during this period, there's no timeout applied at all.

Which is just ... totally broken. Unless I'm misunderstanding? Sigh.

Applying the read timeout to both is more sensible than applying it to neither. I guess there's also an argument for applying the connect timeout to both? But the read timeout is the one that normally acts as a "something has to happen at least this often" timeout, and that's what I'd expect the write timeout to do...

Reply:
I'm using the master branch in my answer below, because at least I have not added mistakes there. :) (But I don't think that makes any difference.)

My understanding is that the timeout sent to `_get_conn` and `LifoQueue.get` is the pool timeout (the time to get a new connection from the queue). When _get_conn returns a connection, `conn.timeout` is set to `timeout_obj.connect_timeout`. (So that even if we reuse a connection, we use the correct timeout.)

How can we end up with a connect timeout that's twice the specific connect timeout? I think we just have pool_timeout + connect_timeout. But pool_timeout is None by default anyway.

And yeah, there's no timeout when we make the request, only when we start reading from it, which is broken. It looks like it's just not possible with http.client: HTTPConnection.request() does not have a timeout argument. So that's something we should try to fix, I guess. Mentioned it in https://github.com/python-trio/urllib3/issues/46.

And yeah, I'm happy to apply the read timeout to both. Merging, thanks.